kubetest/boskos.go

```diff
@@ -0,0 +1,134 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+func renew(job string) error {
+	resp, err := http.Get(fmt.Sprintf("http://boskos/request?job=%v&duration=5m", job))
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("Status %s, StatusCode %v", resp.Status, resp.StatusCode)
+	}
+	return nil
+}
+
+func returnProj(job string) error {
+	resp, err := http.Get(fmt.Sprintf("http://boskos/return?job=%v", job))
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("Status %s, StatusCode %v", resp.Status, resp.StatusCode)
+	}
+	return nil
+}
+
+func requestProj(job string) (string, error) {
+	resp, err := http.Get(fmt.Sprintf("http://boskos/request?job=%v&duration=5m", job))
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == 204 {
+		log.Print("No available projects, retrying in 30sec.")
+		time.Sleep(30 * time.Second)
+		return "", nil
+	} else if resp.StatusCode == 200 {
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return "", err
+		}
+
+		type GCPProject struct {
+			// This need to be consistent with boskos
+			Name     string        `json:"name"`
+			Owner    string        `json:"owner"`
+			Duration time.Duration `json:"duration"`
+			Start    time.Time     `json:"start"`
+		}
+		var p = new(GCPProject)
+		err = json.Unmarshal(body, &p)
+		if err != nil {
+			return "", err
+		}
+		return p.Name, nil
+	} else {
+		return "", fmt.Errorf("Status %s, StatusCode %v", resp.Status, resp.StatusCode)
+	}
+}
+
+func boskos(job string, lease time.Duration) (string, error) {
+	if job == "" {
+		return "", fmt.Errorf("job must not be empty")
+	}
+
+	var proj string
+	for i := 0; i < 3; i++ {
+		proj, err := requestProj(job)
+		if err != nil {
+			return "", err
+		}
+		if proj != "" {
+			break
+		}
+		return "", fmt.Errorf("No available projects")
+	}
+
+	go func(job string, lease time.Duration) {
+		sigs := make(chan os.Signal, 1)
+		leaseChan := time.NewTimer(lease).C
+		tickChan := time.NewTicker(time.Minute * 4).C // Give 1 min overhead
+
+		signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+		for {
+			select {
+			case <-tickChan:
+				log.Print("Renew lease for another 5min")
+				renew(job)
+			case <-leaseChan:
+				log.Print("Lease timed up")
+				returnProj(job)
+				return
+			case <-sigs:
+				log.Printf("Terminated : sig %v", sigs)
+				returnProj(job)
+				return
+			}
+		}
+	}(job, lease)
+
+	return proj, nil
```

Comment:
Right now there is no clean way to use this library to return the lease aside from sending a signal.